### ISSUE TITLE
CP-SAT Solver: Job Shop Problem (JSP) with "strange" results (v7.8.7959 / Windows / Java)

### ISSUE THREAD

**Issue by tachikaw** [google/or-tools#2148]
I'm modeling a railway rescheduling (to solve conflicts in the railway network) as a Job Shop Problem (JSP):

	- Each service / train is a "job", that will be executed by a set of sections of the railway network.
	- Each path of service / train is a "task", that will be executed in a section.
	- Each path of railway network is a "machine", that will "perform" different jobs (services / trains) for a period of time.
	
Based on the example given in "The Job Shop Problem" (https://developers.google.com/optimization/scheduling/job_shop), I added some precedence restrictions in each path of railway network where two or more services / trains pass:  
The services that are scheduled to be executed first have higher precedence, to avoid that in case of conflict of services, a service that comes first has to wait for what comes after, to pass (weird situation).

	model.addGreaterOrEqual( IntVar of task1 service start, IntVar of task0 service end)... 
	model.addGreaterOrEqual( IntVar of task2 service start, IntVar of task1 service end)... 
	...
	
At any given moment, I take a real-time "photograph" of the rail network, with the approximate positions of each service that is running at that time, applying any delays in the remaining scheduled route of each service, using this data as input to the JSP, but before running the JSP, I am able to check conflicts on each path, comparing the timestamps of each service that will go through that path.	

I did some tests, using "clean" schedules (without conflicts) and then introducing delays in some services to create conflicts and
check the robustness of the JSP's results in resolving conflicts.

I'm not going to put code here, which is basically the example of "The Job Shop Problem", plus the rules of precedence that I described above. Instead, I will put some prints that better show the case' contexts and the results.

Case 1: Clean schedule (without conflicts).

	Pre-processing:

		...
		
		Service: 15641
		
		...
		
		servicePathsData.get(4).get(0): 9439172_9402006 (path)
		**servicePathsData.get(4).get(1): 1597770720 (start timestamp)**
		servicePathsData.get(4).get(2): 1597770960 (end timestamp)
		servicePathsData.get(4).get(3): 240 (duration)
		start: GAIA-GEN. TORRES
		end: PORTO-CAMPANHÃ

		...

	JSP runs...

	Post-processing:	

		...
		
		Service: 15641
		
		...

		(path): 9439172_9402006
		**(start timestamp): 1597770840**
		(end timestamp): 1597771080
		(duration): 240
		start: GAIA-GEN. TORRES
		end: PORTO-CAMPANHÃ

		...

	Questions:
		- Why the JSP add 120 s. to the start timestamp of this service' path, even with the "clean" schedule?  

Case 2: Adding delays in schedule (causing conflicts).

	Pre-processing:

		...

		path: 9408102_9408136 [S.MART. DO CAMPO / TERRONHAS]
		pathServicesData.get(0).get(0): 15421 (service)
		pathServicesData.get(0).get(1): 1597770840 (start timestamp): Tue Aug 18 18:14:00 BST 2020
		pathServicesData.get(0).get(2): 180 (duration)
		**Exists CONFLICT between 15421 and 15541**
		pathServicesData.get(1).get(0): 15541 (service)
		pathServicesData.get(1).get(1): 1597770945 (start timestamp): Tue Aug 18 18:15:45 BST 2020
		pathServicesData.get(1).get(2): 105 (duration)
		pathServicesData.get(2).get(0): 15423 (service)
		pathServicesData.get(2).get(1): 1597771740 (start timestamp): Tue Aug 18 18:29:00 BST 2020
		pathServicesData.get(2).get(2): 180 (duration)

		path: 9408136_9408151 [TERRONHAS / TRANCOSO]
		pathServicesData.get(0).get(0): 15421 (service)
		pathServicesData.get(0).get(1): 1597771020 (start timestamp): Tue Aug 18 18:17:00 BST 2020
		pathServicesData.get(0).get(2): 180 (duration)
		**Exists CONFLICT between 15421 and 15541**
		pathServicesData.get(1).get(0): 15541 (service)
		pathServicesData.get(1).get(1): 1597771050 (start timestamp): Tue Aug 18 18:17:30 BST 2020
		pathServicesData.get(1).get(2): 105 (duration)
		pathServicesData.get(2).get(0): 15423 (service)
		pathServicesData.get(2).get(1): 1597771920 (start timestamp): Tue Aug 18 18:32:00 BST 2020
		pathServicesData.get(2).get(2): 180 (duration)

		...

		Conflicts:

			**Conflict in path: 9408136_9408151 [TERRONHAS / TRANCOSO] between services: 15421 and 15541**
			**Conflict in path: 9408102_9408136 [S.MART. DO CAMPO / TERRONHAS] between services: 15421 and 15541**
			Conflict in path: 9408177_9408201 [RECAREI-SOBREIRA / PARADA] between services: 15541 and 15421
			Conflict in path: 9408151_9408177 [TRANCOSO / RECAREI-SOBREIRA] between services: 15541 and 15421

		...

	JSP runs...

		...

		Solve status: OPTIMAL

		Statistics:
		CpSolverResponse:
		status: OPTIMAL
		objective: 3240
		best_bound: 3240
		booleans: 70
		**conflicts: 0**
		branches: 70
		propagations: 0
		integer_propagations: 157
		walltime: 0.0339646
		usertime: 0.0339672
		deterministic_time: 5.606e-06
		primal_integral: 0

		...

	Post-processing:	

		...

		path: 9408102_9408136 [S.MART. DO CAMPO / TERRONHAS]
		service 15421: 1597770840 (start)
		service 15421: 1597771020 (end)
		service 15421: 180 (duration)
		**Exists CONFLICT between 15421 and 15541**
		service 15541: 1597770945 (start)
		service 15541: 1597771050 (end)
		service 15541: 105 (duration)
		service 15423: 1597771740 (start)
		service 15423: 1597771920 (end)
		service 15423: 180 (duration)

		path: 9408136_9408151 [TERRONHAS / TRANCOSO]
		service 15421: 1597771020 (start)
		service 15421: 1597771200 (end)
		service 15421: 180 (duration)
		**Exists CONFLICT between 15421 and 15541**
		service 15541: 1597771050 (start)
		service 15541: 1597771155 (end)
		service 15541: 105 (duration)
		service 15423: 1597771920 (start)
		service 15423: 1597772100 (end)
		service 15423: 180 (duration)

		...

	Questions:
		- Why did the JSP resolve some of the conflicts and not even "touch" the others?		

In short, the modeling of the problem is wrong (is there a new rule missing?) rr should I consider that they are normal limitations
CP-SAT Solver?
What am I doing wrong?

Thanks!

**Comment by lperron** [comment]
Issues have very limited audience.

Please use the mailing list.

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00
